internal/ext/ext.go

```diff
@@ -0,0 +1,104 @@
+package ext
+
+import (
+	"regexp"
+	"fmt"
+)
+
+const (
+	E_TAG_NOT_SUPPORT = "E_TAG_NOT_SUPPORT"
+	E_BAD_TAG	= "E_BAD_TAG"
+)
+
+var MAX_TAG_LEN = 100
+var validTagFmt = regexp.MustCompile(`^[a-zA-Z0-9]+$`)
+
+
+var EXT_VER_NO_EXT = ExtVerNoExt{ExtVer(0x1)}
+var EXT_VER_TAG = ExtVerTag{ExtVer(0x2)}
+
+
+type ExtVerNoExt struct {
+	version ExtVer
+}
+
+func (self *ExtVerNoExt) Version() ExtVer {
+	return self.version
+}
+
+//ext version to support tag filter
+type ExtVerTag struct {
+	version	ExtVer
+}
+
+func (self *ExtVerTag) Version() ExtVer {
+	return self.version
+}
+
+type ExtVer uint8
+
+var NO_EXT_VER = ExtVer(uint8(0x1))
+var NO_EXT = &NoExtContent{}
+
+type Tag []byte
+
+type NoExtContent struct {
+
+}
+
+func (self *NoExtContent) ExtVersion() ExtVer {
+	return NO_EXT_VER
+}
+
+func (self *NoExtContent) GetBytes() []byte {
+	return []byte("")
+}
+
+func NewTag(tagName []byte) (Tag, error) {
+	if tagName != nil && string(tagName) == "" {
+		return nil, nil
+	}
+	if !validateTag(tagName) {
+		return nil, fmt.Errorf("invalid tag %v", tagName)
+	}
+	return Tag(tagName), nil
+}
+
+func (tag Tag) ExtVersion() ExtVer {
+	return ExtVer(uint8(0x2))
+}
+
+func (tag Tag) Match(anotherTag Tag) bool {
+	tagStr := string(tag)
```

Comment:
As client without tag(or we say tag is nil) could receive tagged message. However I think tag is strictly match, if there is no client could receive one specified tag, that message goes to default message channel,  and there is no need to take nil tag into consideration